pkg/resource/api.go

```diff
@@ -184,24 +186,26 @@ func (a *APIUpdatingApplicator) Apply(ctx context.Context, o runtime.Object, ao
 		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
 	}
 
-	current := o.DeepCopyObject()
+	desired := o.DeepCopyObject()
 
-	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, current)
+	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
 	if kerrors.IsNotFound(err) {
 		// TODO(negz): Apply ApplyOptions here too?
-		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+		return errors.Wrap(a.client.Create(ctx, desired), "cannot create object")
 	}
 	if err != nil {
 		return errors.Wrap(err, "cannot get object")
 	}
 
 	for _, fn := range ao {
-		if err := fn(ctx, current, o); err != nil {
+		if err := fn(ctx, o, desired); err != nil {
 			return err
 		}
 	}
 
-	return errors.Wrap(a.client.Update(ctx, o), "cannot update object")
+	desired.(metav1.Object).SetResourceVersion(o.(metav1.Object).GetResourceVersion())
```

Comment:
Could you talk about why we need to do this? From my reading of the code, this will wipe out everything in the remote resource, including `status`, and change it with the template we have. Then the controller in the remote cluster will reconcile and fill it up again; continuous battle of controllers. In fact, since we erase `spec`, there could be leak of actual resources.

In `Service` case, the IP is gone in the next reconcile; the LB controller might think that this resource is new and create a new LB resource without even deleting the old one. In case of our managed resource, `crossplane.io/external-name` will be wiped out if it's not given in template and the controller will recreate a new external resource in each reconcile.

IMO, `Update` should usually be the first choice over `Patch` if you have the `struct` of the target resource and can operate on it instead of dumping a static template. Since here we dump a static template, I think it should always use `Patch`. Even though `Patch` has its own problems, regarding lists, it's better than to encounter late-init problems only in arrays compared to whole metadata, spec, status.

Reply:
I think there is a case for two kinds of applicator:

1. Patching. We overlay our desired state on actual state, preserving any actual state that wasn't included in our desired state.
1. Updating. We replace all actual state with our desired state.

I agree that it wouldn't make sense for to use this updating applicator in many places. KARs for example, should not use this to update their remote resources for the reasons you point out. However I do think there are some cases where we want this:

1. Between KA and KAR. On each reconcile of a KA we want to render out all of its KARs using the templates in the KA. Nothing else ever writes to the spec or metadata of those KARs - it's safe to blow that all away on every reconcile. (We may need to consider KAR status though - I don't think we want to blow that away on each reconcile).
1. Propagated connection secrets. We don't care what the secret we're propagating to looks like. We actually _don't want to_ preserve fields that aren't in the desired state, we want to straight up overwrite the "to" secret with the "from" secret every time.

Reply:
> We may need to consider KAR status though

@muvaf @hasheddan I forget - does calling `client.Update(ctx, o)` with an empty status overwrite the status of the resource in the API server even if it's using the status subresource? If that's the case then this APIUpdatingApplicator is probably a lot less useful.

Reply:
> Between KA and KAR. On each reconcile of a KA we want to render out all of its KARs using the templates in the KA. Nothing else ever writes to the spec or metadata of those KARs - it's safe to blow that all away on every reconcile. (We may need to consider KAR status though - I don't think we want to blow that away on each reconcile).

I am not sure. It's the way things are right now that KAR is a resource whose metadata and spec don't change but that's only an assumption of today. For example, we might want to store a reference of the resource in the remote cluster in `KAR.spec`. It'll be gone in the next reconcile.

> Propagated connection secrets. We don't care what the secret we're propagating to looks like. We actually don't want to preserve fields that aren't in the desired state, we want to straight up overwrite the "to" secret with the "from" secret every time.

No, there are creation time information in those secrets that should be preserved because there is no way to fetch them from the provider again. I remember we were doing get->fill->update but now we rely on `Patch` behavior to keep the old keys intact.

I don't know, it really sounds like a bold assumption to make. It also renders any other controller who'd like to make changes on the CR useless.

I'd say that we should just never set `resourceVersion` if the change is not done on the retrieved object. If necessary, we can investigate doing the patch in-place through iterating over template and setting the values on the retrieved resource.

> does calling client.Update(ctx, o) with an empty status overwrite the status of the resource in the API server even if it's using the status subresource?

I don't think it does. I included `status` in my argument because first I though this will be used to update the resources in the remote cluster, which we don't know whether they have `status` as subresource.

Reply:
> I am not sure. It's the way things are right now that KAR is a resource whose metadata and spec don't change but that's only an assumption of today. For example, we might want to store a reference of the resource in the remote cluster in KAR.spec. It'll be gone in the next reconcile.

Shouldn't we worry about that if and when that becomes a use case?

> No, there are creation time information in those secrets that should be preserved because there is no way to fetch them from the provider again.

I think we're talking about different things. I'm discussing `pkg/reconciler/secret` and the managed resource -> claim secret propagation code. I think you're talking about the code that writes the "source of truth" managed resource secrets, not the various places we replicate those to.

Reply:
In my mind what we're doing here with update is not dramatically more dangerous than what we do in a patch. In either case we take desired state and apply it without concern as to the resource version of an existing resource. This seems reasonable to me as an option _as long as the caller knows what they're doing_. I'm not suggesting this replace Patch everywhere, or become the new default. Just that there seem to be cases where it could be useful.

Reply:
I feel like doing this makes more sense than the current implementation as the default with the current will always fail unless you have already gotten the current, in which case you might as well call Update yourself. If we think this new behavior is too dangerous I would prefer to still change it to at least call Update with the object that we get so the default is a no-op rather than an error. That would make it more resemble CreateOrUpdate and you would be expected to pass in some sort of mutation in an ApplyOption. However, I think with appropriate warnings the behavior this PR implements is potentially useful.

Reply:
> Shouldn't we worry about that if and when that becomes a use case?

In general yes, but I think this is a case where we break something until a use case comes up rather than over-abstraction for future.

> I think we're talking about different things. I'm discussing pkg/reconciler/secret and the managed resource -> claim secret propagation code. I think you're talking about the code that writes the "source of truth" managed resource secrets, not the various places we replicate those to.

Yeah, I was thinking about managed -> claim.

> In my mind what we're doing here with update is not dramatically more dangerous than what we do in a patch. In either case we take desired state and apply it without concern as to the resource version of an existing resource. This seems reasonable to me as an option as long as the caller knows what they're doing.

> That would make it more resemble CreateOrUpdate and you would be expected to pass in some sort of mutation in an ApplyOption.

If they want that specifically, I think it's dangerous enough that it should be an `ApplyOption` for them to opt-in explicitly rather than doing it in all cases. By default, it will simply be like `CreateOrUpdate`, i.e. useful in cases where they first `Get` -> if NotFound, then dump from template, if Found then do some changes -> `Apply`. Maybe an option like `ForceUpdate()` or `OverrideResourceVersion()`.

Reply:
Yeah we had talked about that pattern as well. If we go that route I think I would prefer us still change to Update with the object that we Get and then in the force override just set that object to the one we passed in but with the resource version. That way the default with no apply option is the no-op.

Reply:
Any new thoughts here @negz @muvaf? I think I would be fine with the most recent proposal here.

Reply:
@hasheddan we (@muvaf and I) just discussed this out of band and we're okay moving forward with the most recent proposal. Thanks for bearing with us!